src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+
+    // If we have already issued an async read
+    read_in_progress: bool,
+
+    // If we received a BROKEN_PIPE or other error
+    // indicating that the remote end has closed the pipe
+    closed: bool,
+
+    // this is ONLY here for debugging purposes.
+    // (for printfs so we know which iocp this is part of)
+    iocp: HANDLE,
+}
+
+impl MessageReader {
+    fn new(handle: HANDLE) -> MessageReader {
+        MessageReader {
+            handle: handle,
+            ov: create_overlapped(),
+            read_buf: Vec::with_capacity(READ_BUFFER_SIZE),
+            read_in_progress: false,
+            closed: false,
+            iocp: 0 as HANDLE,
+        }
+    }
+
+    fn ov_ptr(&self) -> *mut winapi::OVERLAPPED {
+        self.ov.deref() as *const _ as *mut winapi::OVERLAPPED
+    }
+
+    fn reset_overlapped(&mut self) {
+        reset_overlapped(&mut self.ov);
+    }
+
+    fn cancel_io(&mut self) {
+        unsafe {
+            if self.read_in_progress {
+                kernel32::CancelIoEx(self.handle, self.ov_ptr());
+                self.read_in_progress = false;
+            }
+        }
+    }
+
+    fn ensure_is_reading(&mut self) -> Result<(),WinError> {
+        self.start_read()
+    }
+
+    // Called when we receive an IO Completion Packet for this handle.
+    fn notify_completion(&mut self, err: u32) -> Result<(),WinError> {
+        dd2!("[$ {:?}:{:?}] notify_completion", self.iocp, self.handle);
+
+        if err == winapi::ERROR_BROKEN_PIPE {
+            assert!(!self.closed, "we shouldn't get an async BROKEN_PIPE after we already got one");
+            self.closed = true;
+            return Ok(());
+        }
+
+        let nbytes = self.ov.InternalHigh as u32;
+        let offset = self.ov.Offset;
+
+        assert!(offset == 0);
+
+        // if the remote end closed...
+        if err != winapi::ERROR_SUCCESS {
+            // This should never happen
+            panic!("[$ {:?}:{:?}] *** notify_completion: unhandled error reported! {}", self.iocp, self.handle, err);
+        }
+
+        unsafe {
+            let new_size = self.read_buf.len() + nbytes as usize;
+            dd!("nbytes: {}, offset {}, buf len {}->{}, capacity {}",
+                nbytes, offset, self.read_buf.len(), new_size, self.read_buf.capacity());
+            assert!(new_size <= self.read_buf.capacity());
+            self.read_buf.set_len(new_size);
+        }
+
+        Ok(())
+    }
+
+    // kick off an asynchronous read
+    fn start_read(&mut self) -> Result<(),WinError> {
+        if self.read_in_progress || self.closed {
+            return Ok(());
+        }
+
+        dd2!("[$ {:?}:{:?}] start_read ov {:?}", self.iocp, self.handle, self.ov_ptr());
+
+        let buf_len = self.read_buf.len();
+        let mut buf_cap = self.read_buf.capacity();
+        let mut bytes_read: u32 = 0;
+
+        // if the buffer is full, add more capacity
+        if buf_cap == buf_len {
+            self.read_buf.reserve(match buf_cap {
```

Comment:
Not sure I follow. AIUI, when you have a large message, you send the message header (containing the data size) and the OOB header (including the dedicated channel handle) in the main message; so before you start receiving on the dedicated channel, you should already have all the information you need? Is this just about the fact that this information is presently not passed down to the reader object in the receiver implementation?...